Mathlib/RingTheory/IsPrimary.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.LinearAlgebra.Quotient
+import Mathlib.RingTheory.Ideal.Operations
+
+/-!
+# Primary submodules
+
+A proper submodule `S : Submodule R M` is primary iff
+  `r • x ∈ S` implies `x ∈ S` or `∃ n : ℕ, r ^ n • (⊤ : Submodule R M) ≤ S`.
+
+## Main results
+
+* `Submodule.isPrimary_iff_zero_divisor_quotient_imp_nilpotent_smul`:
+  A `N : Submodule R M` is primary if any zero divisor on `M ⧸ N` is nilpotent.
+  See `https://mathoverflow.net/questions/3910/primary-decomposition-for-modules`
+  for a comparison of this definition (a la Atiyah-Macdonald) vs "locally nilpotent" (Matsumura).
+
+## Implementation details
+
+This is a generalization of `Ideal.IsPrimary`. For brevity, the pointwise instances are used
+to define the nilpotency of `r : R`.
+
+-/
+
+open Pointwise
+
+namespace Submodule
+
+section CommSemiring
+
+variable {R M : Type*} [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/-- A proper submodule `S : Submodule R M` is primary iff
+  `r • x ∈ S` implies `x ∈ S` or `∃ n : ℕ, r ^ n • (⊤ : Submodule R M) ≤ S`.
+  This generalizes `Ideal.IsPrimary`. -/
+def IsPrimary (S : Submodule R M) : Prop :=
```

Comment:
Consider making this `protected`, and perhaps the other as well (in a separate PR, if it has much fallout)

Reply:
Most current references of `IsPrimary` for ideals looks like `IsPrimary I` so making `Ideal.IsPrimary` protected would require changing several files.